### ISSUE TITLE
Angular v9 Support

### ISSUE THREAD

**Issue by puku0x** [OnsenUI/OnsenUI#2723]
Finally, Angular v9 has been published! 🎉
https://blog.angular.io/version-9-of-angular-now-available-project-ivy-has-arrived-23c97b63cfa3

I think it is also a good opportunity to switch from `gulp + webpack` to [ng-packagr](https://github.com/ng-packagr/ng-packagr) to make the future migration easy.

**Comment by emccorson** [comment]
Thank you for your suggestions. On the face of it, both of these seem like good ideas.

Let me look into it in more detail next week so I can understand what is involved, and let's discuss it further then.

**Comment by emccorson** [comment]
I'm trying to wrap my head around how the switch to ng-packagr might work - I was wondering if you have any thoughts?:

Firstly, is there a difference between building with Angular CLI and building ng-packagr. As far as I can see, both have the same build output so I'm guessing ng-packagr is part of Angular CLI. Is that correct?

The other thing is the structure of Angular libraries generated with Angular CLI. As far as I can tell, you always have a workspace, and inside the workspace is `projects/my-library`.

```
.
└── bindings
    └── angular2    <--- workspace/default project, maybe move example application here?
        └── projects
            └── ngx-onsenui    <--- library project, source files here
```

So to make Onsen UI conform to the directory structure of Angular CLI, we would need to move the source files into `projects/ngx-onsenui` or something like that.

That leaves the question of what application should be in the workspace. Maybe it would be useful to move the example application in `bindings/angular2/examples` to the default application.

I think this should conform to the correct Angular CLI directory structure.

What do you think? @puku0x

**Comment by puku0x** [comment]
@emccorson
Yes, ng-packagr is now part of Angular CLI. I think `ng g lib ngx-onsenui` makes it easy. And I agree with the idea keeping the example applications in `bindings/angular2/examples` for smooth transition.

Talking about the directory structure, it is also possible to have no `projects` directory because the path to source files is configurable in `angular.json`. Here is an example: https://github.com/puku0x/ngx-onsenui-ng-packagr-test

**Comment by emccorson** [comment]
@puku0x 
Thank you for clarifying the directory structure question and thank for creating an example repository.

Yesterday I was reading more about what is required and I think I understand what to do now. I'll be working on moving the Angular bindings to Angular CLI this week. I will link the PR here when it is ready.

**Comment by emccorson** [comment]
I have been working more on the Angular 9 + CLI work, but I am getting stuck on an error when building the schematics.

```
Building schematics...

schematics/src/ng-add/setup-project.ts:133:49 - error TS2345: Argument of type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/typescript/lib/typescript").SourceFile' is not assignable to parameter of type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript").SourceFile'.

133     const changes = addSymbolToNgModuleMetadata(moduleSource, appModulePath, 'schemas', 'CUSTOM_ELEMENTS_SCHEMA', '@angular/core');
                                                    ~~~~~~~~~~~~

schematics/src/utils/index.ts:53:37 - error TS2345: Argument of type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/typescript/lib/typescript").SourceFile' is not assignable to parameter of type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript").SourceFile'.
  Types of property 'kind' are incompatible.
    Type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/typescript/lib/typescript").SyntaxKind.SourceFile' is not assignable to type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript").SyntaxKind.SourceFile'.

53   const changes = addImportToModule(moduleSource, modulePath, moduleName, src);
                                       ~~~~~~~~~~~~
```

Do you know why the error is appearing?

If you have time, I would really appreciate if you could try running the build to see if you also get the error. I set up a new repository for holding the work temporarily:

**Comment by emccorson** [comment]
@puku0x I have been working more on the Angular 9 + CLI work, but I am getting stuck on an error when building the schematics.

```
Building schematics...

schematics/src/ng-add/setup-project.ts:133:49 - error TS2345: Argument of type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/typescript/lib/typescript").SourceFile' is not assignable to parameter of type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript").SourceFile'.

133     const changes = addSymbolToNgModuleMetadata(moduleSource, appModulePath, 'schemas', 'CUSTOM_ELEMENTS_SCHEMA', '@angular/core');
                                                    ~~~~~~~~~~~~

schematics/src/utils/index.ts:53:37 - error TS2345: Argument of type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/typescript/lib/typescript").SourceFile' is not assignable to parameter of type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript").SourceFile'.
  Types of property 'kind' are incompatible.
    Type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/typescript/lib/typescript").SyntaxKind.SourceFile' is not assignable to type 'import("/home/eric/src/OnsenUI/bindings/mangular2/node_modules/@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript").SyntaxKind.SourceFile'.

53   const changes = addImportToModule(moduleSource, modulePath, moduleName, src);
                                       ~~~~~~~~~~~~
```

Do you know why the error is appearing? My guess is that the same library is being imported two different ways and the typechecker can't figure out that they are the same library.

If you have time, I would really appreciate if you could try running the build to see if you also get the error. I set up a new repository for holding the work temporarily:
https://github.com/emccorson/mangular2

To run it with Onsen UI, these are the instructions:

```
git clone git@github.com:emccorson/mangular2
rm -rf OnsenUI/bindings/angular2
cp -r mangular2 OnsenUI/bindings/angular2
cd OnsenUI/bindings/angular2
ng build
cd projects/ngx-onsenui
npm run build
```

Once you do that it should try to build the schematics but fail.

**Comment by puku0x** [comment]
@emccorson I see. I will take a look.

**Comment by puku0x** [comment]
@emccorson 
It seems the error message says the version of TypeScript is different between `@schematics/*` and `ngx-onsenui`.  I could build the schematics after I updated the version of TypeScript in `bindings/angular2/package.json` to `"typescript": "~3.8.0"` and re-install dependencies.

**Comment by emccorson** [comment]
@puku0x Your fix worked. Thank you! That helped me a lot.

**Comment by tahir-jamil** [comment]
With angular 9 i feel multiple issues 
1- mockStatus bar function not called needs to be tested
2- navigator with angular 9 not worked properly i have added multiple issues about this too 
3- OnsLazyRepeat does not support lvy


Angular 9 is very very fast for large scale applications and with Onsen its a new era of mobile development with pwa's 
i hope we will get angular 9 complete support asap.

[https://github.com/OnsenUI/OnsenUI/issues/2754](https://github.com/OnsenUI/OnsenUI/issues/2754)

[https://github.com/OnsenUI/OnsenUI/issues/2753](https://github.com/OnsenUI/OnsenUI/issues/2753)

[https://github.com/OnsenUI/OnsenUI/issues/2752](https://github.com/OnsenUI/OnsenUI/issues/2752)

[https://github.com/OnsenUI/OnsenUI/issues/2751](https://github.com/OnsenUI/OnsenUI/issues/2751)

**Comment by jkon** [comment]
So, has angular > v8 support been abandoned? otherwise I need to replace onsen for mobile support in my app very soon

**Comment by emccorson** [comment]
It is still on the list of things to do. I attempted it before but ran into
trouble. Currently I am working on the upgrade to support Vue 3 and intend
to try the Angular upgrade again after that.

**Comment by jkon** [comment]
Having been trying to upgrade past angular 8 for some time myself, I understand. But I feel like I'm at a point I have to do it, so I have stripped out onsen and replaced it with another framework. I will still follow onsen and try it again if you get it upgraded

**Comment by freon27** [comment]
I know that the answer is "make a PR", but is Angular support basically dead unless someone comes along who wants to support this?

**Comment by emccorson** [comment]
It has been a long time since the Angular bindings were upgraded and we are behind, but it is still on the list of things to get done and is not forgotten about. A PR would be very much appreciated from someone who is interested in seeing the Angular bindings upgraded and we would assist as much as possible, but it will eventually get done.

**Comment by freon27** [comment]
Thanks for the response - that's good to know. 

I don't know if it's something I will get time to work on but at least if i know it's still a goal for the project, it's worth considering it.

**Comment by freon27** [comment]
And would the plan be to skip straight to the current version rather than providing incremental support for 9-13?

**Comment by emccorson** [comment]
Thank you for your consideration. Yes, I think the most likely route would be to skip straight to the latest version.

**Comment by emccorson** [comment]
It has been a long time since the Angular bindings were upgraded and we are behind, but it is still on the list of things to get done and is not forgotten about. A PR would be very much appreciated from someone who is interested in seeing the Angular bindings upgraded and we would assist as much as possible, but it will eventually get done.
